### DISCUSSION TITLE
Cannot process DELETE request via CORS

### DISCUSSION THREAD

**Discussion by libovness** [vercel/next.js#24778] [Help]
### What version of Next.js are you using?

10.1.3

### What version of Node.js are you using?

15.11.0

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

Vercel (but I'm running it locally)

### Describe the Bug

I can process `GET` and `POST` via CORS, but cannot process `DELETE`, at least not locally. It's not even reaching my handler, so the middleware solution discussed [here](https://nextjs.org/docs/api-routes/api-middlewares) does not help.

One thing that I don't understand, and this may be he response to the `DELETE` request only includes GET and OPTIONS: 
`Access-Control-Allow-Methods: OPTIONS, GET`

### Expected Behavior

`DELETE` request reaches my handler

### To Reproduce

Add this to `next.config.js`

```
async headers() {
  return [
    {
      source: "/api/:path*",
      headers: [
        { key: "Access-Control-Allow-Credentials", value: "true" },
        { key: "Access-Control-Allow-Origin", value: "*" },
        { key: "Access-Control-Allow-Methods", value: "GET,OPTIONS,PATCH,DELETE,POST,PUT" },
        { key: "Access-Control-Allow-Headers", value: "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version" },
      ]}
  ]
}
```

Here's my api handler: 

```
import { makeUserSupabase } from '../../../../../../../../lib/initSupabase'
import Cors from 'cors'

// I added this per https://nextjs.org/docs/api-routes/api-middlewares but it makes no difference because my request is not even reaching the handler 
const cors = Cors({
  methods: ['GET', 'HEAD','DELETE'],
})

// Helper method to wait for a middleware to execute before continuing
// And to throw an error when an error happens in a middleware
function runMiddleware(req, res, fn) {
  return new Promise((resolve, reject) => {
    fn(req, res, (result) => {
      if (result instanceof Error) {
        return reject(result)
      }

      return resolve(result)
    })
  })
}

async function deleteSavedTemplate(req, res) {

  /* Code */
    
}

const main = async (req, res) => {

  await runMiddleware(req, res, cors)

  // Never logs
  console.log('we have arrived')  
  if (req.method === 'DELETE') {
    return await deleteSavedTemplate(req, res)
  } else {
    return res.status(405).end();
  }

}

export default main
```

I'm calling it like this in `/pages/index.js`
```
const headers = new Headers({
   ...supabase.storage.headers,
  'Content-Type': 'text/plain'
 })

const url = `${apiServer}/v1/users/me/templates/popular/saves/${template.id}`
  await fetch(url, {
    headers: headers,
    method: 'DELETE'              
 })
```
Where ${apiServer} is either http://localhost:8000 or our domain

**Comment by krychaxp** [comment]
i dont know if its help, but you can use https://www.npmjs.com/package/next-connect

**Comment by libovness** [comment]
thanks @krychaxp but I don't believe this will work as the request isn't even hitting my handler

**Comment by timneutkens** [comment]
Just tried this exact reproduction and the handler is being called correctly both from the same domain and other domains (so the preflight is handled correctly). Even `deleteSavedTemplate` is called (added a `console.log` there).

**Reply by libovness** [reply]
huh

this even worked when both were served on localhost?

**Reply by timneutkens** [reply]
Yeah

**Reply by libovness** [reply]
I neglected to include the full next.config.js, which, the more I think about it, seems like the likely culprit, as the CORS and methods headers clearly aren't coming back in the preflight

I know this isn't a bug now but if anyone can spare some kindness, this is what my `next.config.js` looks like.

```
module.exports = {
  async redirects() {
       return [
          {
            source: '/api/:dest*',
            destination: '/:dest*',
            permanent: true,
          }
      ]
    },  
    async rewrites() {
      return [
          {
            source: "/v1/:dest*",
            destination: "/api/v1/:dest*",         
        },
        {
            source: "/v0/:dest*",
            destination: "/api/v0/:dest*",        
        }
      ]
    },
  async headers() {
    return [
      {
        // matching all API routes
        source: "/api/:dest*",
        headers: [
          { key: "Access-Control-Allow-Credentials", value: "true" },
          { key: "Access-Control-Allow-Origin", value: "*" },
          { key: "Access-Control-Allow-Methods", value: "GET,OPTIONS,PATCH,DELETE,POST,PUT" },
          { key: "Access-Control-Allow-Headers", value: "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version" },
        ]
      }
    ]
  }
};
```

**Reply by richardpickett** [reply]
It's been a while since the above problem was posted, likely it's not around anymore, but I'd like to share a little from my experience with CORS, nextjs, and next-auth.

There are soooooo many factors that can be causing this problem.  I recommend reading my write-up over on the next-auth repo, especially start at the bottom to read a recap and then read the longer solution further up.

https://github.com/nextauthjs/next-auth/issues/4327
